prusti-interface/src/environment/mod.rs

```diff
@@ -277,36 +287,96 @@ impl<'tcx> Environment<'tcx> {
-                // that was used to store the data.
-                let body_with_facts = unsafe {
-                    self::mir_storage::retrieve_mir_body(self.tcx, def_id)
-                };
-                let body = body_with_facts.body;
-                let facts = BorrowckFacts {
-                    input_facts: RefCell::new(Some(body_with_facts.input_facts)),
-                    output_facts: body_with_facts.output_facts,
-                    location_table: RefCell::new(Some(body_with_facts.location_table)),
-                };
-
-                CachedBody {
-                    base_body: Rc::new(body),
-                    monomorphised_bodies: HashMap::new(),
-                    borrowck_facts: Rc::new(facts),
-                }
-            });
-        body
-            .monomorphised_bodies
+        let body = if let Some(body) = pure_function_bodies.get_mut(&def_id) {
+            body
+        } else if let Some(body) = predicate_bodies.get_mut(&def_id) {
+            body
+        } else {
+            // fallback to fetching the non-pure mir body, which may not be cached yet
+            let body = bodies.entry(def_id)
+                .or_insert_with(|| {
+                    self.raw_load_local_mir(def_id)
+                });
+            body
+        };
+
+        self.subst_into_body(&mut body.body, substs)
+    }
+
+    // used only internally
+    fn external_spec_mir(&self, def_id: DefId, substs: SubstsRef<'tcx>) -> Rc<mir::Body<'tcx>> {
+        let mut external_spec_bodies = self.external_spec_bodies.borrow_mut();
+        let body = external_spec_bodies.get_mut(&def_id)
+            .unwrap_or_else(|| panic!("External body of spec {:?} was not imported!", def_id));
+        self.subst_into_body(body, substs)
+    }
+
+    /// Get the MIR body of a spec, monomorphised with the given type substitutions.
+    /// Can be both local or non-local to the current crate.
+    pub fn spec_mir(&self, def_id: DefId, substs: SubstsRef<'tcx>) -> Rc<mir::Body<'tcx>> {
+        if let Some(def_id) = def_id.as_local() {
+            let mut local_spec_bodies = self.local_spec_bodies.borrow_mut();
+            let mut pure_function_bodies = self.pure_function_bodies.borrow_mut();
+            let mut predicate_bodies = self.predicate_bodies.borrow_mut();
+            let body = if let Some(body) = local_spec_bodies.get_mut(&def_id) {
+                Some(body)
+            } else if let Some(body) = pure_function_bodies.get_mut(&def_id) {
+                Some(body)
+            } else {
+                predicate_bodies.get_mut(&def_id)
+            }.unwrap_or_else(|| panic!("Local body of spec {:?} was not loaded with `load_local_spec_mir`!", def_id));
+            self.subst_into_body(&mut body.body, substs)
+        } else {
+            self.external_spec_mir(def_id, substs)
+        }
+    }
+
+    /// Similar to `spec_mir` but also allows lookup in local bodies.
```

Comment:
I'm not sure I understand this comment. Doesn't this method perform fewer lookups, not more?